engine/schema/src/main/java/com/cloud/upgrade/dao/DatabaseAccessObject.java

```diff
@@ -84,6 +85,31 @@ public boolean columnExists(Connection conn, String tableName, String columnName
         return columnExists;
     }
 
+    public void addIndexIfNeeded(Connection conn, String tableName, String columnName) {
```

Comment:
this can be extracted to 2 atom operations
- indexExists
- createIndex

Reply:
Yeah, I was thinking about that but then I have to create a separate index naming function, etc. I guess I was just being lazy.

Reply:
Before I do this rework, looking for feedback. Do we want to:

1) create index naming function to generate index name
2) call indexExists with tableName and indexName
3) call createIndex with tableName and indexName

It seems to me we would prefer to keep the index naming scheme private to outside callers so it isn't something overridable and no potential for mismatches between checking existence and creating - which leads me to think we would want to retain a single entry point function like the `addIndexIfNeeded` to generate the name privately and then call `indexExists` and `createIndex` if we still want to break these out.

Reply:
agree with you
- add `indexExists` and `createIndex` in `DatabaseAccessObject`
- add `addIndexIfNeeded` in `DbUpgradeUtils`, which calls `indexExists` and `createIndex` 
- call `DbUpgradeUtils.addIndexIfNeeded` in Upgrade*.java

Reply:
Cool, see update @weizhouapache. Tested this as well.